metricbeat/beater/reload.go

```diff
@@ -0,0 +1,221 @@
+package beater
+
+import (
+	"expvar"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/publisher"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+var (
+	debugr        = logp.MakeDebug("reloader")
+	configReloads = expvar.NewInt("metricbeat.config.reloads")
+	moduleStarts  = expvar.NewInt("metricbeat.config.module.starts")
+	moduleStops   = expvar.NewInt("metricbeat.config.module.stops")
+	moduleRunning = expvar.NewInt("metricbeat.config.module.running")
+
+	ErrNoNewFiles = errors.New("no new files found")
+)
+
+type ConfigReloader struct {
+	registry RunningRegistry
+	config   ModulesReloadConfig
+	client   func() publisher.Client
+	done     chan struct{}
+	wg       sync.WaitGroup
+}
+
+type RunningRegistry struct {
+	List  map[uint64]ModuleRunner
+	mutex sync.Mutex
```

Comment:
If the mutex is required to access `List` then they should have the same visibility. But I like embedding the mutex in the struct then you can call `runningReg.Lock()`.